cosmwasm/packages/registry/src/registry.rs

```diff
@@ -323,66 +344,121 @@ impl<'a> Registry<'a> {
     ///
     /// `block_time` is the current block time. This is needed to calculate the
     /// timeout timestamp.
+    #[allow(clippy::too_many_arguments)]
     pub fn unwrap_coin_into(
         &self,
         coin: Coin,
-        receiver: String,
+        receiver_addr: String,
         into_chain: Option<&str>,
         own_addr: String,
         block_time: Timestamp,
-        with_memo: String,
+        first_transfer_memo: String,
+        receiver_callback: Option<Callback>,
     ) -> Result<proto::MsgTransfer, RegistryError> {
-        let path = self.unwrap_denom_path(&coin.denom)?;
+        // Calculate the path that this coin took to get to the current chain.
+        // Each element in the path is an IBC hop.
+        let path: Vec<MultiHopDenom> = self.unwrap_denom_path(&coin.denom)?;
         self.deps
             .api
             .debug(&format!("Generating unwrap transfer message for: {path:?}"));
 
-        let MultiHopDenom {
-            local_denom: _,
-            on: first_chain,
-            via: first_channel,
-        } = path
-            .first()
-            .ok_or_else(|| RegistryError::InvalidDenomTracePath {
-                path: format!("{:?}", path.clone()),
-                denom: coin.denom.clone(),
-            })?;
+        // Validate Path
+        if path.is_empty() {
+            return Err(RegistryError::InvalidDenomTracePath {
+                path: format!("{path:?}"),
+                denom: coin.denom,
+            });
+        }
+
+        // Define useful variables
+        let current_chain = path[0].on.clone();
+        let first_channel = path[0].via.clone();
+
+        let coin_type = if path.len() == 1 {
+            CoinType::Native
+        } else {
+            CoinType::Ibc
+        };
```

Comment:
Still not clear on what were doing for paths >= 3. Is that for a future PR to return an error here and rewind?

Reply:
We can build a memo for unwinding paths of any length. So a path >=3 will just do more hops before being sent to the destination. See this test, for example: https://github.com/osmosis-labs/osmosis/blob/main/tests/ibc-hooks/ibc_middleware_test.go#L933

Reply:
Oh awesome!